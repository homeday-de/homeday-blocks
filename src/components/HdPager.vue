<template>
  <div
    v-if="count > 1"
    :class="{
      pager: true,
      isRippleEffectActive: hasRippleEffect,
      isUsingMouse: isUsingMouse,
      isWhite: white,
    }"
    @keydown="setUsingMouse(false)"
    @mousedown="setUsingMouse(true)"
  >
    <nav
      :style="{
        height: `${dotSize}px`,
        width: `${maxVisible * dotSize}px`,
      }"
      class="pager__items"
      tabindex="0"
      @keydown.stop.self="maybeSelectItem"
      @blur="setUsingMouse(false)"
    >
      <ul
        :style="{
          transform: `translateX(${trackOffset}px)`,
        }"
        class="pager__items__track"
      >
        <div
          :style="{
            'flex-basis': `${outOfRangeItemsOffset}px`,
          }"
          class="pager__items__offset"
        />
        <li
          v-for="item in items"
          :key="item.value"
          :class="{
            'pager__items__item': true,
            [`pager__items__item--size-${item.size}`]: true,
            isActive: item.active,
          }"
          :style="{
            height: `${dotSize}px`,
            width: `${dotSize}px`,
          }"
          :aria-current="item.active ? 'true' : 'false'"
          @click="triggerRippleEffect(); $emit('input', item.value)"
        >
          <span class="pager__items__item__text">
            {{ item.value + 1 }}
          </span>
        </li>
      </ul>
    </nav>
  </div>
</template>

<script>
import { getArrayOfSize } from 'homeday-blocks/src/services/utils';

const DOT_SIZE = 40;
let rippleTimeout;

export default {
  name: 'HdPager',
  props: {
    value: {
      type: Number,
      default: 0,
      validator(value) {
        return value >= 0;
      },
    },
    count: {
      type: Number,
      default: 1,
      validator(count) {
        return count >= 1;
      },
    },
    maxVisible: {
      type: Number,
      default: 7,
      validator(maxVisible) {
        return maxVisible > 3;
      },
    },
    white: {
      type: Boolean,
      default: false,
    },
  },
  data() {
    return {
      dotSize: DOT_SIZE,
      hasRippleEffect: false,
      trackOffset: this.count > this.maxVisible
        ? DOT_SIZE
        : 0,
      isUsingMouse: false,
    };
  },
  computed: {
    outOfRangeItemsOffset() {
      return getArrayOfSize(this.count).reduce((acc, current) => {
        const paddingCount = this.maxVisible - 1;
        const position = current;
        if (position < (this.value + (paddingCount * -1))) {
          return acc + 1;
        }
        return acc;
      }, 0) * DOT_SIZE;
    },
    items() {
      return getArrayOfSize(this.count).reduce((acc, current) => {
        const paddingCount = this.maxVisible - 1;
        const position = current;
        if (position < (this.value + (paddingCount * -1))
        || position > this.value + this.maxVisible + paddingCount) {
          // This item is out of range
          return acc;
        }
        const newItem = {
          value: current,
          position,
          offset: position * DOT_SIZE,
          size: this.getSize(position),
          active: current === this.value,
        };
        return [...acc, newItem];
      }, []);
    },
  },
  watch: {
    value() {
      this.$nextTick(() => {
        this.calculateTrackOffset();
      });
    },
    maxVisible() {
      this.$nextTick(() => {
        this.calculateTrackOffset();
      });
    },
    count() {
      this.$nextTick(() => {
        this.calculateTrackOffset();
      });
    },
  },
  methods: {
    getSize(position) {
      if (this.count <= this.maxVisible) {
        return 'regular';
      }
      if (this.maxVisible === 3) {
        return 'small';
      }
      const offset = position * DOT_SIZE;
      const isFirstVisible = (offset + this.trackOffset) === 0;
      const isLastVisible = (offset + this.trackOffset)
        - ((this.maxVisible - 1) * DOT_SIZE) === 0;
      if (isFirstVisible || isLastVisible) {
        return 'tiny';
      }
      if (this.maxVisible < 7) {
        return 'regular';
      }
      const isImmediatelyAfterFirstVisible = (offset + this.trackOffset) === DOT_SIZE;
      const isImmediatelyBeforeLastVisible = (offset + this.trackOffset)
        - ((this.maxVisible - 1) * DOT_SIZE) === DOT_SIZE * -1;
      if (this.isOutOfBoundsLeft(position) || this.isOutOfBoundsRight(position)) {
        return 'tiny';
      }
      if (isImmediatelyAfterFirstVisible || isImmediatelyBeforeLastVisible) {
        return 'small';
      }
      return 'regular';
    },
    isOutOfBoundsLeft(position) {
      const offset = position * DOT_SIZE;
      return offset + this.trackOffset < 0;
    },
    isOutOfBoundsRight(position) {
      const offset = position * DOT_SIZE;
      return (offset + this.trackOffset) - ((this.maxVisible - 1) * DOT_SIZE) > 0;
    },
    triggerRippleEffect() {
      this.hasRippleEffect = false;
      this.$nextTick(() => {
        this.hasRippleEffect = true;
        clearTimeout(rippleTimeout);
        rippleTimeout = setTimeout(() => {
          this.hasRippleEffect = false;
        }, 550);
      });
    },
    calculateTrackOffset() {
      if (this.count < this.maxVisible) {
        this.trackOffset = 0;
        return;
      }
      const currentOffset = this.trackOffset;
      const valueOffset = this.value * DOT_SIZE;
      const isFirstVisible = (valueOffset + currentOffset) === 0;
      const isLastVisible = (valueOffset + currentOffset)
        - ((this.maxVisible - 1) * DOT_SIZE) === 0;
      if (isFirstVisible) {
        this.trackOffset += DOT_SIZE;
        return;
      }
      if (isLastVisible) {
        this.trackOffset -= DOT_SIZE;
      }
      if (valueOffset + currentOffset < 0) {
        this.trackOffset = (valueOffset * -1) + DOT_SIZE;
      }
      if ((valueOffset + currentOffset) - ((this.maxVisible - 1) * DOT_SIZE) > 0) {
        this.trackOffset = Math.max(
          (valueOffset - ((this.maxVisible - 2) * DOT_SIZE)) * -1,
          ((this.count * DOT_SIZE) - ((this.maxVisible - 1) * DOT_SIZE)) * -1,
        );
      }
    },
    maybeSelectItem(e) {
      if (['ArrowDown', 'ArrowRight'].includes(e.key)) {
        const currentIndex = this.value;
        if (currentIndex === this.count - 1) {
          this.$emit('input', 0);
          return;
        }
        this.$emit('input', currentIndex + 1);
      } else if (['ArrowUp', 'ArrowLeft'].includes(e.key)) {
        const currentIndex = this.value;
        if (currentIndex === 0) {
          this.$emit('input', this.count - 1);
          return;
        }
        this.$emit('input', currentIndex - 1);
      }
    },
    setUsingMouse(usingMouse) {
      this.isUsingMouse = usingMouse;
    },
  },
};
</script>

<style lang="scss" scoped>
@import 'homeday-blocks/src/styles/mixins.scss';
@import 'homeday-blocks/src/styles/inputs.scss';

@keyframes pager-ripple {
  0% {
    opacity: 1;
  }
  20% {
    transform: translate(-50%, -50%) scale((20 / 8));
  }
  100% {
    opacity: 0;
  }
}
.pager {
  $_root: &;
  display: flex;
  justify-content: center;
  &__items {
    display: flex;
    justify-content: center;
    position: relative;
    overflow: hidden;
    transition: outline 0.1s ease-in-out;
    #{$_root}.isUsingMouse & {
      outline: 0;
    }
    &__track {
      display: flex;
      align-items: center;
      flex: 0 1 auto;
      max-width: 100%;
      transition: transform ($time-s * 2) ease-in-out;
    }
    &__offset {
      flex-shrink: 0;
    }
    &__item {
      position: relative;
      flex-grow: 0;
      flex-shrink: 0;
      cursor: pointer;
      user-select: none;
      -webkit-tap-highlight-color: transparent;
      &::before {
        position: absolute;
        top: 50%;
        left: 50%;
        transform: translate(-50%, -50%);
        display: block;
        content: '';
        background-color: getShade($quaternary-color, 60);
        border-radius: 50%;
        transition:
          height $time-s ease-in-out,
          width $time-s ease-in-out;
        z-index: 2;

        #{$_root}.isWhite & {
          background-color: $primary-bg;
        }
      }
      &.isActive::before {
        background-color: getShade($secondary-color, 110);
      }
      &--size-regular::before {
        width: 8px;
        height: 8px;
      }
      &--size-small::before {
        width: 6px;
        height: 6px;
      }
      &--size-tiny::before {
        width: 4px;
        height: 4px;
      }
      &--size-regular.isActive::before,
      &--size-small.isActive::before,
      &--size-tiny.isActive::before,
      &--size-regular:hover::before,
      &--size-small:hover::before,
      &--size-tiny:hover::before {
        width: 12px;
        height: 12px;
        background-color: getShade($secondary-color, 110);
      }
      &::after {
        position: absolute;
        top: 50%;
        left: 50%;
        width: 8px;
        height: 8px;
        transform: translate(-50%, -50%);
        display: block;
        content: '';
        background-color: getShade($secondary-color, 80);
        border-radius: 50%;
        opacity: 0;

        #{$_root}.isWhite & {
          background-color: rgba(white, 0.3);
        }
      }
      &.isActive::after,
      #{$_root}.isRippleEffectActive &.isActive::after {
        animation: 500ms pager-ripple ease-in-out;
      }
      &__text {
        border: 0;
        clip: rect(1px, 1px, 1px, 1px);
        clip-path: inset(50%);
        height: 1px;
        margin: -1px;
        overflow: hidden;
        padding: 0;
        position: absolute;
        width: 1px;
        word-wrap: normal !important;
      }
    }
  }
}
</style>
