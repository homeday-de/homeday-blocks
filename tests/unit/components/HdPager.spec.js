import { wrapperFactoryBuilder } from 'tests/unit/helpers';
import HdPager from '@/components/HdPager.vue';

const wrapperBuilder = wrapperFactoryBuilder(HdPager, {
  props: {
    count: 10,
  },
});

describe('HdPager', () => {
  // list of classNames we depend on
  const PAGER_ITEMS_SELECTOR = '.pager__items';
  const PAGER_ITEM_SELECTOR = '.pager__items__item';
  const ACTIVE_PAGER_ITEM_SELECTOR = `${PAGER_ITEM_SELECTOR}.isActive`;

  it('The component is rendered', () => {
    const wrapper = wrapperBuilder();

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('Pre-selecting the active item should work', () => {
    const wrapper = wrapperBuilder({
      props: {
        value: 3,
      },
    });

    // We increment the nth-child by 1 because there is always a sizer item
    // before the pager items
    expect(wrapper.findAll(`${ACTIVE_PAGER_ITEM_SELECTOR}:nth-child(5)`).length).toBe(1);
  });

  it('By default, first item is selected', () => {
    const wrapper = wrapperBuilder();

    // We increment the nth-child by 1 because there is always a sizer item
    // before the pager items
    expect(wrapper.findAll(`${ACTIVE_PAGER_ITEM_SELECTOR}:nth-child(2)`).length).toBe(1);
  });

  it('Pager item can be selected', () => {
    const wrapper = wrapperBuilder();

    // select a pager item
    wrapper.find(`${PAGER_ITEM_SELECTOR}:nth-child(5)`).trigger('click');
    // confirm that input event is emitted
    expect(wrapper.emitted('input')).toBeTruthy();
    // And that expected payload is passed with the event
    const payload = wrapper.emitted('input')[0][0];
    expect(payload).toBe(3);
  });

  it('Pager supports keyboard navigation', () => {
    const wrapper = wrapperBuilder({
      props: {
        value: 3,
      },
    });

    const pagerItems = wrapper.find(PAGER_ITEMS_SELECTOR);
    let payload;
    // Left arrow
    pagerItems.trigger('keydown.left', { key: 'ArrowLeft' });
    expect(wrapper.emitted('input')).toBeTruthy();
    // eslint-disable-next-line prefer-destructuring
    payload = wrapper.emitted('input')[0][0];
    expect(payload).toBe(2);
    // Up arrow
    pagerItems.trigger('keydown.up', { key: 'ArrowUp' });
    expect(wrapper.emitted('input')).toBeTruthy();
    // eslint-disable-next-line prefer-destructuring
    payload = wrapper.emitted('input')[1][0];
    expect(payload).toBe(2);
    // Right arrow
    pagerItems.trigger('keydown.right', { key: 'ArrowRight' });
    expect(wrapper.emitted('input')).toBeTruthy();
    // eslint-disable-next-line prefer-destructuring
    payload = wrapper.emitted('input')[2][0];
    expect(payload).toBe(4);
    // Down arrow
    pagerItems.trigger('keydown.down', { key: 'ArrowDown' });
    expect(wrapper.emitted('input')).toBeTruthy();
    // eslint-disable-next-line prefer-destructuring
    payload = wrapper.emitted('input')[3][0];
    expect(payload).toBe(4);
  });
});
