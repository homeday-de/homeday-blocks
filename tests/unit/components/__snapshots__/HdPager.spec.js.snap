// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HdPager The component is rendered 1`] = `
<div class="pager">
  <nav tabindex="0" class="pager__items" style="height: 40px; width: 280px;">
    <ul class="pager__items__track" style="transform: translateX(40px);">
      <div class="pager__items__offset" style="flex-basis: 0px;"></div>
      <li aria-current="true" class="pager__items__item pager__items__item--size-small isActive" style="height: 40px; width: 40px;"><span class="pager__items__item__text">
          1
        </span></li>
      <li aria-current="false" class="pager__items__item pager__items__item--size-regular" style="height: 40px; width: 40px;"><span class="pager__items__item__text">
          2
        </span></li>
      <li aria-current="false" class="pager__items__item pager__items__item--size-regular" style="height: 40px; width: 40px;"><span class="pager__items__item__text">
          3
        </span></li>
      <li aria-current="false" class="pager__items__item pager__items__item--size-regular" style="height: 40px; width: 40px;"><span class="pager__items__item__text">
          4
        </span></li>
      <li aria-current="false" class="pager__items__item pager__items__item--size-small" style="height: 40px; width: 40px;"><span class="pager__items__item__text">
          5
        </span></li>
      <li aria-current="false" class="pager__items__item pager__items__item--size-tiny" style="height: 40px; width: 40px;"><span class="pager__items__item__text">
          6
        </span></li>
      <li aria-current="false" class="pager__items__item pager__items__item--size-tiny" style="height: 40px; width: 40px;"><span class="pager__items__item__text">
          7
        </span></li>
      <li aria-current="false" class="pager__items__item pager__items__item--size-tiny" style="height: 40px; width: 40px;"><span class="pager__items__item__text">
          8
        </span></li>
      <li aria-current="false" class="pager__items__item pager__items__item--size-tiny" style="height: 40px; width: 40px;"><span class="pager__items__item__text">
          9
        </span></li>
      <li aria-current="false" class="pager__items__item pager__items__item--size-tiny" style="height: 40px; width: 40px;"><span class="pager__items__item__text">
          10
        </span></li>
    </ul>
  </nav>
</div>
`;
